### ISSUE TITLE
Join requires tables to be specified in sorted order

### ISSUE THREAD

**Issue by nathanielc** [influxdata/flux#4521]
This code works:

```
import "generate"

rando = generate.from(
    count: 6,
    fn: (n) => (n + 1) * (n + 2),
    start: 2021-01-01T00:00:00Z,
    stop: 2021-01-02T00:00:00Z,
)

t1 = rando
    |> group()

t2 = t1
    |> group()

merged = join(
        tables: {t1: t1, t2: t2},
        on: ["_time"]
    )

merged |> yield()
```

This code panics:

```import "generate"

rando = generate.from(
    count: 6,
    fn: (n) => (n + 1) * (n + 2),
    start: 2021-01-01T00:00:00Z,
    stop: 2021-01-02T00:00:00Z,
)

t1 = rando
    |> group()

t2 = t1
    |> group()

merged = join(
        //this is the only change
        tables: {t1: t1, d2: t2},
        on: ["_time"]
    )

merged |> yield()
```

Panic:
```
panic: Inconsistent plan graph: successor does not have edge back to predecessor"}

goroutine 16670 [running]:

runtime/debug.Stack()

/home/circleci/.tools/go/src/runtime/debug/stack.go:24 +0x65

github.com/influxdata/influxdb/v2/query/control.(*Controller).executeQuery.func1()

/home/circleci/go/src/github.com/influxdata/influxdb/query/control/controller.go:457 +0x125

panic({0x3b69c00, 0x400bc48})

/home/circleci/.tools/go/src/runtime/panic.go:1038 +0x215

github.com/influxdata/flux/plan.updateSuccessors(0xc00c0ea640, {0x40f0768, 0xc01e730fc0}, {0x40f07f0, 0xc008a15d60})

/home/circleci/go/pkg/mod/github.com/influxdata/flux@v0.139.0/plan/heuristic_planner.go:170 +0x305

github.com/influxdata/flux/plan.(*heuristicPlanner).Plan(0x4002398, {0x40a6bc0, 0xc021ec7e00}, 0xc00c0ea640)

/home/circleci/go/pkg/mod/github.com/influxdata/flux@v0.139.0/plan/heuristic_planner.go:118 +0x374

github.com/influxdata/flux/plan.(*physicalPlanner).Plan(0xc00aacb728, {0x40a6bc0, 0xc021ec7e00}, 0x406cab8)

/home/circleci/go/pkg/mod/github.com/influxdata/flux@v0.139.0/plan/physical.go:47 +0x4c

github.com/influxdata/flux/plan.(*planner).Plan(0xc00c0ad3c0, {0x40a6bc0, 0xc021ec7e00}, 0x0)

/home/circleci/go/pkg/mod/github.com/influxdata/flux@v0.139.0/plan/builder.go:49 +0x8a

github.com/influxdata/flux/lang.buildPlan({0x40a6bc0, 0xc021ec7e00}, 0xc022fd8150, 0xc01fa7d270)

/home/circleci/go/pkg/mod/github.com/influxdata/flux@v0.139.0/lang/compiler.go:157 +0x34d

github.com/influxdata/flux/lang.(*AstProgram).Start(0xc01fa7d950, {0x40a6bc0, 0xc00a71bfb0}, 0xc001385f6c)

/home/circleci/go/pkg/mod/github.com/influxdata/flux@v0.139.0/lang/compiler.go:500 +0x570

github.com/influxdata/influxdb/v2/query/control.(*Controller).executeQuery(0xc001385f78, 0xc00b518480)

/home/circleci/go/src/github.com/influxdata/influxdb/query/control/controller.go:480 +0x226

github.com/influxdata/influxdb/v2/query/control.(*Controller).processQueryQueue(0xc000fc5800)

/home/circleci/go/src/github.com/influxdata/influxdb/query/control/controller.go:438 +0x2e

github.com/influxdata/influxdb/v2/query/control.New.func1()

/home/circleci/go/src/github.com/influxdata/influxdb/query/control/controller.go:220 +0x5d

created by github.com/influxdata/influxdb/v2/query/control.New

/home/circleci/go/src/github.com/influxdata/influxdb/query/control/controller.go:218 +0xa1b
```

**Comment by skartikey** [comment]
The issue doesn't seem inside the Join function. 

Below works - 
```
import "generate"

rando1 = generate.from(
    count: 6,
    fn: (n) => (n + 1) * (n + 2),
    start: 2021-01-01T00:00:00Z,
    stop: 2021-01-02T00:00:00Z,
)

rando2 = generate.from(
    count: 6,
    fn: (n) => (n + 1) * (n + 2),
    start: 2021-01-01T00:00:00Z,
    stop: 2021-01-02T00:00:00Z,
)

t1 = rando1
    |> group()

t2 = rando2
    |> group()

merged = join(
        tables: {t1: t1, d2: t2},
        on: ["_time"]
    )

merged |> yield()
```

It seems the issue is inside the planner, if the table is not specified in the sorted order the successor node is not able to find the edge back to the predecessor. https://github.com/influxdata/flux/blob/master/plan%2Fheuristic_planner.go#L160-L172

```
t1 = rando
    |> group()

t2 = t1
    |> group()
```
The above part of the query is creating some sort of circular dependency. More investigation is needed.

**Comment by wolffcm** [comment]
This error message indicates that there is a bug in how to plan graph is being constructed. I don't see how any rules would fire for this script, otherwise I would say a rule could be corrupting the plan.

The plan graph is doubly linked, each node has a set of pointers to all of its successors and also all of its predecessors. As such we expect a reciprocal relationship there. If a plan node A has plan node B listed among its successors, then plan node B must have plan node A listed among its predecessors.

The problem must be with how we construct the initial plan, which is done here:
https://github.com/influxdata/flux/blob/3651f5e2422f9cce9a33817b87e77bc63e7b7e90/plan/logical.go#L170

I'm guessing that because `join` is somewhat special in that it has a variable number of predecessors, our code is not robust to that somewhere.

**Comment by wolffcm** [comment]
This was fixed by #4853
